### ISSUE TITLE
Cannot create complex circular relationships between tables and datatypes

### ISSUE THREAD

**Issue by ghost** [yesodweb/persistent#147]
The title here is a bit misleading, but the basic idea is that complex mutually-dependent tables and datatypes are not currently possible.

An example:

```
Foo
    name Text
Bar
    name Text
    parent Parent
```

This schema will work fine if Parent is declared before the splice occurs. However, we run into a problem if we want the following:

``` haskell
data Parent = ParentFoo FooId | ParentBar BarId
    deriving (Show, Read, Eq)
```

This datatype depends on the datatypes introduced by `mkPersist`, but `mkPersist` also depends on those datatypes. The natural (for an experienced programmer, anyway) approach is to include the declaration of `Parent` in the splice:

``` haskell
$(concat <$> sequence [ [d| data Parent = ParentFoo $(conT $ mkName "FooId")
                                        | ParentBar $(conT $ mkName "BarId")
                        |]
                      , derivePersistField "Parent"
                      , mkPersist ps $(persistFileWith lowerCaseSettings modelFile)
                      ])
```

This approach still has a problem though! Because `persistFileWith` returns the `SqlType` unadorned in the `FieldDef`, the compiler attempts to splice in the `Parent` datatype in the inner splice, which naturally fails.

This two-stage problem has been seen before, as is evidenced by the code in `getSqlType` in `Database.Persist.TH`, which has to work around the issue.

Better, I think, would be to return a lifted `Exp`, changing the type of the `EntityDef` spliced in by `persistFileWith` (and other functions that produce `EntityDef`s) from `EntityDef SqlType` to `EntityDef SqlTypeExp`, as this will allow the splicing of the `SqlType` to be delayed until the outer splice, where the new datatype is visible and able to be compiled mutually with the rest of the database schema.

**Comment by snoyberg** [comment]
Closing out old issues, please reopen if still relevant

**Comment by scshunt** [comment]
This is still an issue. I've come across it again in designing an auth system like http://blog.felipe.lessa.nom.br/?p=7 where I want auth information stored in the database, so I thought about it again.

The failings were almost certainly caused by the fact that other parts of the code were expecting to do other things with the EntityDefs, and this predictably didn't work with the delayed lift.

In order to preserve the option of the existing behaviour, there are a couple of options available. One option, to which I'm most partial, is to add a new `persistWithDelayed` (or `persistWithThunk` or `persistWithExp` or any better name someone comes up with) which creates, rather than `[EntityDef]`, a `[EntityDefExp]`. 

We then add a new typeclass, say `LiftEntityDef`, and it has a method `liftEntityDef`. `EntityDef` will have an instance where `liftEntityDef = lift`, and `EntityDefExp` will have an instance where `liftEntityDef = id`. All of the TH functions expecting a `[EntityDef]` currently can be amended to expect a value of this class, allowing them to behave the same whether or not we use the thunked persist or not. To a user, this should be entirely transparent, and preserves the old behaviour of being able to generate an `[EntityDef]` directly.

**Comment by creichert** [comment]
Reopening for  further discussion. I'm curious if the changes mentioned earlier in the thread were actually merged from the persistent2 branch.

**Comment by gregwebs** [comment]
persistent2 branch changes were merged which improved the situation a lot.

@scshunt ultimately is your goal to define a data type outside of persistent? I think it is essential for persistent to support that use case but I have not found the time to be able to implement it.
I prototyped something where the user can add annotations to an existing data type: those annotations are then available to TH at compile time.

**Comment by scshunt** [comment]
The goal is to define a data type which is mutually recursive to the data types defined by persistent inside the call to `mkPersist`.

Based on what I've seen, the technical obstacle remains the same even after those changes. Currently, the archetypical usage involves using a `persistWith`-family function to generate a list of `EntityDef`s, which are then fed into other TH functions to generate the declarations, as in the following example from the book:

    share [mkPersist sqlSettings, mkMigrate "migrateAll"] [persistLowerCase|
    Person
        name String
        age Int Maybe
        deriving Show
    BlogPost
        title String
        authorId PersonId
        deriving Show
    |]

The issue arises because there are actually two splices here. The first is that the quasi-quoter generates a `Q Exp` which is spliced in to produce the `[EntityDefs]` object; I'll call this the inner splice. The same occurs when using `persistFileWith`, except that in this case the splice is done explicitly. The second is the outer splice where the `Q [Dec]` results of `share` are spliced in to declare everything.

The field types (such as `Parent`) in the original example are referred to in the inner splice, specifically in the `fieldSqlType` field of the corresponding `FieldDef`. For instance:

        parent Parent

will have a `fieldSqlType` of `sqlType (Proxy Parent)`.

This produces an issue if `Parent` is not yet declared, which can't happen if we want `Parent` to be mutually recursive with the types declared by the `mkPersist` call. This leads to my suggested solution. In the common use-case where the `[EntityDef]` created by `persistWith` is consumed only by TH functions, we do not actually need to produce a `[EntityDef]` containing the `SqlType`, since that type isn't used by `mkPersist` and other functions (or, at least, it wasn't when I conceived of the previous patch). This means we don't actually need that type when we do the inner splice, and we can save it for the outer splice.

This opens the door to mutually recursive definitions. If we splice the definition of `Parent` in at the same time as we splice the results of `mkPersist`, the resulting definitions can freely refer to each other.

**Comment by gregwebs** [comment]
persistent2 helped this out by relying on laziness. 8646be3ebde181cd3e17334a14c38d47c5503772 broke any cycles.

@scshunt please go forward with your plan of attack

**Comment by scshunt** [comment]
Sounds good. I will develop a proof-of-concept, and then you guys can tell me whether existing facilities (like `EmbedFieldDef` and the like, which I just learned exist) can be reused/combined in a way to make this easier to work with.

**Comment by scshunt** [comment]
Ok... so what I'm seeing now is that this approach should work, but doing so will require replicating `FieldDef`'s and `EntityDef`'s structure I believe. In my earlier code, I took advantage of the fact that they were parametric on the SQL type, but this is no longer the case. I'm not entirely sure what the purpose of delaying the lifts in `EntityDefSqlTypeExp` is for, either... is this an attempt to solve this problem? Or a related problem? In any case I would like to expand upon the documentation there to make it more clear why they are necessary.

I'm not sure of the best approach... one that seems viable would be to go with `data EntityDef' a = ...`, and then do `type EntityDef = EntityDef' SqlType` and only use the latter in public-facing code, except for in the template library. In there, we would use something like `EntityDef Exp` (typed TH would be nice eventually...) and have a class---I'm thinking of calling it `ReifySqlType a`---so that `mkPersist` and friends change their parameter from `[EntityDef]` to `(ReifySqlType a) => [EntityDef' a]`.

I will add a parameter to `EntityDefSqlTypeExp` and related types, to indicate whether or not to reify the `fieldSqlType` on the first lift, or to delay the lift even further and produce an `EntityDef Exp` instead, leaving the reification for another step.

Thoughts? Does this seem right?

**Comment by gregwebs** [comment]
> In any case I would like to expand upon the documentation there to make it more clear why they are necessary.

Essentially when parsing we try to go straight into the EntityDef and FieldDef structure which is a mistake and led to the recessionary parametrization which leaked out into the rest of persistent. I was able to deal with that and other edge cases by adding wrappers like `EntityDefSqlTypeExp` which deals with the possibility of `fieldSqlType` being an expresdynsion, which should not be a concern of `FieldDef`.

As another potential approach, I would be all for getting rid of the fieldSqltype from `FieldDef` because it is too sql specific. It could either be typeclassed, moved into the `fieldAttrs`, or we figure out some other new way to separate out migration information.

**Comment by scshunt** [comment]
Ah, yes, I would like that. Why can `fieldSqlType` be different from the `sqlType` of the underlying type anyway?

**Comment by gregwebs** [comment]
A user can specify an arbitrary `sqltype=`. Fundamentally we want the flexibility to have variations on how a  Haskell data type may be treated in the database without having to always newtype for that purpose, even if Persistent should have good defaults. But setting sqltype= should really be just a migration concern.

**Comment by NorfairKing** [comment]
+1 still having this issue.

**Comment by parsonsmatt** [comment]
going to close this as fixed by #1256 

if there are remaining use cases that are similar, please post them as separate issues
